icepyx/core/auth.py

```diff
@@ -0,0 +1,147 @@
+            self._s3_initial_ts = datetime.datetime.now()
+            
+        # Only generate s3login_credentials the first time credentials are accessed, or if an hour
+        # has passed since the last login    
+        if self._s3login_credentials is None:
+            set_s3_creds()
+        elif (datetime.datetime.now() - self._s3_initial_ts) >= datetime.timedelta(hours=1):
+            set_s3_creds()
+        return self._s3login_credentials
+
+    def earthdata_login(self, uid=None, email=None, s3token=False, **kwargs) -> None:
+        """
+        Authenticate with NASA Earthdata to enable data ordering and download.
+
+        Generates the needed authentication sessions and tokens, including for cloud access.
+        Authentication is completed using the [earthaccess library](https://nsidc.github.io/earthaccess/).
+        Methods for authenticating are:
+            1. Storing credentials as environment variables ($EARTHDATA_LOGIN and $EARTHDATA_PASSWORD)
+            2. Entering credentials interactively
+            3. Storing credentials in a .netrc file (not recommended for security reasons)
+        More details on using these methods is available in the [earthaccess documentation](https://nsidc.github.io/earthaccess/tutorials/restricted-datasets/#auth).
+        The input parameters listed here are provided for backwards compatibility;
+        before earthaccess existed, icepyx handled authentication and required these inputs.
+        DevNote: Maintained for backward compatibility
+
+        Parameters
+        ----------
+        uid : string, default None
+            Deprecated keyword for Earthdata login user ID.
+        email : string, default None
+            Deprecated keyword for backwards compatibility.
+        s3token : boolean, default False
+            Deprecated keyword to generate AWS s3 ICESat-2 data access credentials
+        kwargs : key:value pairs
+            Keyword arguments to be passed into earthaccess.login().
+
+        Examples
+        --------
+        >>> reg_a = ipx.Query('ATL06',[-55, 68, -48, 71],['2019-02-20','2019-02-28']) # doctest: +SKIP
+        >>> reg_a.earthdata_login() # doctest: +SKIP
+        Enter your Earthdata Login username: ___________________
+
+        EARTHDATA_USERNAME and EARTHDATA_PASSWORD are not set in the current environment, try setting them or use a different strategy (netrc, interactive)
+        No .netrc found in /Users/username
+
+        """
+
+        auth = earthaccess.login(**kwargs)
+        if auth.authenticated:
+            self._auth = auth
+            self._session = auth.get_session()
+
+        if s3token == True:
+            self._s3login_credentials = auth.get_s3_credentials(daac="NSIDC")
+
+        if uid != None or email != None:
+            warnings.warn(
+                "The user id (uid) and/or email keyword arguments are no longer required.",
+                DeprecationWarning,
+            )
```

Comment:
Just noticed in a hackweek tutorial that this warning doesn't show up even if someone supplies email or user id - wondering if it's because of the earthaccess prints resulting from logging in and if putting this check first would make it show up when run in a notebook.

Reply:
I looked into this and it turns out we need to add a `stacklevel=2` argument. [The docs](https://docs.python.org/3/library/warnings.html#warnings.warn) says:
> The stacklevel argument can be used by wrapper functions written in Python ... [which] makes the warning refer to deprecation()’s caller, rather than to the source of deprecation() itself

To be honest I don't understand _exactly_ what that means, but generally it sounds like we need the argument to push the warning all the way up the stack to the user, as right now it's getting stuck somewhere in between the source code and the user.